### ISSUE TITLE
Add uninitialize function in distributed

### ISSUE THREAD

**Issue by zw0610** [pytorch/pytorch#30371]
## 🚀 Feature

Add a function in distributed package to uninitialize the default process group.

## Motivation

We are trying to make a fault-tolerant distributed training mode for pytorch. Or even more generally, the number of workers can be adjusted during the training process without checkpoint. That asks us to rebuild the default process group, which might need to come after the existing process group uninitialized.

We tested with simple NCCL ops and found that creating a second NCCL communicator results an error, which can be remedied by calling [`ncclCommAbort`](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/api/comms.html#ncclcommabort).

## Pitch

After the program realized a loss or join of a worker, the uninitialization function will be called to nullify the existing default process group and function `torch.distributed.init_process_group` will be called for again with a new configuration to create a new default process group. 

## Additional context

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

**Comment by zw0610** [comment]
I do found `destroy_process_group` api in the source code, but wonder why this api is not exposed in doc.

**Comment by pietern** [comment]
There's two steps to achieving this, destroying the process group being the easy one:

* Detect the error: first, you have to figure out there's an error in the first place. Historically, NCCL didn't support this, but with 2.4 introduced `ncclCommAbort`, allowing you to cancel a running kernel. Before you call this, however, you need to know whether there is an error or timeout or such. By default, NCCL kernel launches are asynchronous (meaning you don't actually wait for them to complete), so step 1 is to start waiting for them to complete. You can do this with an undocumented feature that makes waiting on a NCCL operation to complete synchronous, allowing you to actually time out or raise an error. This can be enabled by setting the environment variable `NCCL_BLOCKING_WAIT` to 1. See #25012 for the PR that added it.
* Only then, you can destroy the process group. You can use `destroy_process_group` for this. 

cc @pritamdamania87 @kiukchung
